Fix newly created account and wallet ui display bug

Sirmorrison opened the pull request:
Fix #597
fix #607
This PR 
*updates create wallet modal to mock up ui.*
<img width="565" alt="image" src="https://user-images.githubusercontent.com/27733432/132923490-9cd7c8b8-b6c1-4fcc-ac92-59cf924a634d.png">
- it updates the colors on the password strength progress bar (weak, moderate, strong)
- it add an info message under the spend password editor
- update the button design to Mock up design
- update the modal button colors when all editors are empty or when an editor has input data

*it adds the create account modal*
<img width="538" alt="image" src="https://user-images.githubusercontent.com/27733432/132924017-55d67ca2-272f-4279-8b04-239f2f8afb71.png">

- Fix bug with account not showing up on the ui after creation.
- Set maximum modal width to prevent modal from over stretching when app is in full screen and clean up modal widget
- Fix bug with progress bar filling up to 100% instead of progressing
- Fix bug with wallet list refresh after wallet is deleted or created
- Update info and password model button colors and styles to mock up design.
- Hide modal when backdrop/background is clicked.

Sirmorrison committed:
Fix bug with account not showing up after creation
- add create account modal and clean up code

Sirmorrison committed:
fix issues with modal extending too much on wider screens by adding a fixed width
 - fix issues with wallet not showing up on ui after creation
 - update modal button to mock up and added passward info wallet modal

Sirmorrison committed:
update modal ui
- Add password length info to create wallet modal.
- set weak, moderate and strong password strength color on progress bar
- Fix bug with progress bar filling up to 100% instead of progressing
- Fix bug with list wallet refresh after wallet is deleted
- Update info and password model button colours and styles to mock up design.
- Hide modal when backdrop/background is clicked.

dreacot commented on ui/decredmaterial/modal.go:
```diff
@@ -10,6 +10,7 @@ import (
 
 type Modal struct {
 	overlayColor color.NRGBA
+	co           color.NRGBA
```
What is co? perhaps we can use a more descriptive name here

Sirmorrison committed:
- Hide account selector modal when back drop is clicked
- rename background color variable
- add top and bottom padding to modal content

beansgum commented on ui/modal/info_modal.go:
```diff
@@ -212,8 +223,6 @@ func (in *InfoModal) actionButtonsLayout() layout.Widget {
 					}
 
 					in.btnPositve.Text = in.positiveButtonText
-					in.btnPositve.Background, in.btnPositve.Color = in.Theme.Color.Surface, in.Theme.Color.Primary
```
A default background and color should be set so for pages without explicit positive button styles.

![Screenshot 2021-09-14 at 10 21 32](https://user-images.githubusercontent.com/19960200/133231746-15bc6326-6277-4a15-bf1e-9e9807784c74.png)

beansgum commented on ui/modal/create_watch_only_modal.go:
```diff
@@ -110,6 +110,11 @@ func (cm *CreateWatchOnlyModal) Handle() {
 		}
 	}
+
+	if cm.modal.BackdropClicked() {
```
A check should be made before closing to avoid closing the modal while a process is running.

beansgum commented on ui/modal/create_password_modal.go:
```diff
@@ -148,6 +159,11 @@ func (cm *CreatePasswordModal) Handle() {
 	}
 
+	if cm.modal.BackdropClicked() {
+		cm.Dismiss()
```
A check should be made before closing to avoid closing the modal while a process is running.

beansgum commented on ui/modal/info_modal.go:
```diff
@@ -136,6 +142,11 @@ func (in *InfoModal) Handle() {
 		in.negativeButtonClicked()
 	}
+
+	if in.modal.BackdropClicked() {
```
A `cancelable` property should be added for pages that do not want the modal to be canceled by tapping outside, some pages require a definite Yes or No.

beansgum commented on ui/modal/create_watch_only_modal.go:
```diff
@@ -110,6 +110,11 @@ func (cm *CreateWatchOnlyModal) Handle() {
+
+	if cm.modal.BackdropClicked() {
+		cm.Dismiss()
+		cm.RefreshWindow()
```
This window refresh should be done here instead https://github.com/planetdecred/godcr/blob/master/ui/window.go#L207

beansgum commented on ui/page/wallet_settings_page.go:
```diff
@@ -245,7 +245,7 @@ func (pg *WalletSettingsPage) Handle() {
 							pm.Dismiss()
 							pg.Toast.Notify("Wallet removed")
-							pg.PopFragment()
+							pg.ChangeFragment(NewWalletPage(pg.Load))
```
This will be adding a new wallet page to the stack, PopFragment should be used to go back to the previous wallet page.

beansgum commented on ui/page/wallet_page.go:
```diff
@@ -129,6 +129,10 @@ func (pg *WalletPage) ID() string {
+	pg.loadWalletAndAccounts()
+}
+
+func (pg *WalletPage) loadWalletAndAccounts() {
```
Should probably protect this with a mutex and instances where `pg.listItems` is accessed.

Sirmorrison committed:
Remove create account modal and revert code
- fix send and remove modal display button colors
- fix info and textinput button style issues.
- add mutex to wallet page
- fix issue with Popfragement not calling onresume()
- clean up code.

beansgum commented on ui/modal/create_password_modal.go:
```diff
@@ -46,10 +49,14 @@ func NewCreatePasswordModal(l *load.Load) *CreatePasswordModal {
-	cm.btnPositve.TextSize, cm.btnNegative.TextSize = values.TextSize16, values.TextSize16
-	cm.btnPositve.Font.Weight, cm.btnNegative.Font.Weight = text.Bold, text.Bold
+	cm.btnNegative.TextSize = values.TextSize16
+	cm.btnNegative.Font.Weight = text.Bold
```
Use `text.Medium` for modal buttons according to the mockup.
```suggestion
	cm.btnNegative.Font.Weight = text.Medium
```

beansgum commented on ui/modal/create_password_modal.go:
```diff
@@ -18,24 +19,26 @@ const CreateWallet = "create_wallet_modal"
-	isLoading      bool
-	materialLoader material.LoaderStyle
+	isLoading         bool
+	isMinimizable     bool
```
The modal is actually being dismissed not minimized. 
```suggestion
	isCancelable     bool
```

beansgum commented on ui/modal/create_password_modal.go:
```diff
@@ -184,7 +208,25 @@ func (cm *CreatePasswordModal) Layout(gtx layout.Context) D {
+					return layout.Inset{Left: values.MarginPadding20, Right: values.MarginPadding20}.Layout(gtx, func(gtx C) D {
+						return layout.Flex{Spacing: layout.SpaceBetween}.Layout(gtx,
+							layout.Rigid(func(gtx C) D {
+								txt := cm.Theme.Label(values.MarginPadding14, "This spending password is for the new wallet only")
```
Why was this text added here? It's not included in the mockup and it shows up if I try to setup a startup password in settings.

Sirmorrison replied:
<img width="389" alt="image" src="https://user-images.githubusercontent.com/27733432/133384420-cb25a17b-3115-4d02-8bda-3424a595930a.png">
it is in the mock up..
i will fix the setting page.

beansgum commented on ui/modal/create_watch_only_modal.go:
```diff
@@ -80,6 +84,10 @@ func (cm *CreateWatchOnlyModal) SetLoading(loading bool) {
 	cm.isLoading = loading
 }
 
+func (cm *CreateWatchOnlyModal) MinimizableBackground(min bool) {
```
Can use `Cancelable(min bool)` or `SetCancelable` instead.

beansgum commented on ui/modal/text_input_modal.go:
```diff
@@ -2,28 +2,34 @@ package modal
 
-	IsLoading bool
+	IsLoading     bool
+	ShowWarnInfo  bool
```
Should probably be specific since the info isn't customizable to any text. 
```suggestion
	ShowAccountWarnInfo  bool
```

beansgum commented on ui/page/receive_page.go:
```diff
@@ -345,7 +345,7 @@ func (pg *ReceivePage) Handle() {
 			Title("Receive DCR").
 			Body("Each time you receive a payment, a new address is generated to protect your privacy.").
-			PositiveButton("Got it", func() {})
+			NegativeButton("Got it", func() {})
```
Why is change done in multiple places?

Sirmorrison replied:
The `NegativeButton` is designed to minimize modals when no specific action is to be carried out.

Sirmorrison committed:
Rename variables
- show wallet info on create wallet modal
- clean up code

beansgum commented on ui/page/wallet_page.go:
```diff
@@ -410,6 +425,8 @@ func (pg *WalletPage) layoutOptionsMenu(gtx layout.Context, optionsMenuIndex int
 func (pg *WalletPage) walletSection(gtx layout.Context) layout.Dimensions {
 	return pg.walletsList.Layout(gtx, len(pg.listItems), func(gtx C, i int) D {
+		pg.listLock.Lock()
+		defer pg.listLock.Unlock()
```
Copy into a local variable, this defer could have unintended effects.
```suggestion
func (pg *WalletPage) walletSection(gtx layout.Context) layout.Dimensions {
	pg.listLock.Lock()
	listItems := pg.listItems
	pg.listLock.Unlock()
	return pg.walletsList.Layout(gtx, len(listItems), func(gtx C, i int) D {
```

beansgum commented on ui/page/wallet_page.go:
```diff
@@ -534,6 +551,8 @@ func (pg *WalletPage) watchOnlyWalletSection(gtx layout.Context) layout.Dimensio
 
 func (pg *WalletPage) layoutWatchOnlyWallets(gtx layout.Context) D {
 	return pg.watchWalletsList.Layout(gtx, len(pg.listItems), func(gtx C, i int) D {
+		pg.listLock.Lock()
```
Copy into a local variable.

Sirmorrison committed:
copy Mutex to local variable

beansgum commented on ui/page/wallet_page.go:
```diff
@@ -129,9 +132,16 @@ func (pg *WalletPage) ID() string {
 	wallets := pg.WL.SortedWalletList()
 
+	pg.listLock.Lock()
 	pg.listItems = make([]*walletListItem, 0)
```
Instead of doing this, store the `walletListItem` in a temp local slice and copy to locked slice when done.
```suggestion
	listItems := make([]*walletListItem, 0)
```

beansgum commented on ui/page/wallet_page.go:
```diff
@@ -410,7 +425,11 @@ func (pg *WalletPage) layoutOptionsMenu(gtx layout.Context, optionsMenuIndex int
+		pg.listLock.Lock()
 		listItem := pg.listItems[i]
+		pg.listLock.Unlock()
+
```
Put a local copy outside the loop to avoid access the mutex in every loop run.
```suggestion
func (pg *WalletPage) walletSection(gtx layout.Context) layout.Dimensions {
	pg.listLock.Lock()
	listItems := pg.listItems
	pg.listLock.Unlock()
	return pg.walletsList.Layout(gtx, len(listItems), func(gtx C, i int) D {

		listItem := pg.listItems[i]

```

beansgum commented on ui/page/wallet_page.go:
```diff
@@ -534,7 +553,11 @@ func (pg *WalletPage) watchOnlyWalletSection(gtx layout.Context) layout.Dimensio
 func (pg *WalletPage) layoutWatchOnlyWallets(gtx layout.Context) D {
 	return pg.watchWalletsList.Layout(gtx, len(pg.listItems), func(gtx C, i int) D {
+
+		pg.listLock.Lock()
```
Put a local copy outside the loop to avoid access the mutex in every loop run.

beansgum commented on ui/page/wallet_page.go:
```diff
@@ -169,9 +179,14 @@ func (pg *WalletPage) OnResume() {
 	}
+
+	pg.listLock.Lock()
+	pg.listItems = append(pg.listItems, listItems...)
```
Is append the intended operation here?
```suggestion
	pg.listItems = listItems
```

beansgum commented on ui/page/wallet_page.go:
```diff
@@ -169,9 +179,14 @@ func (pg *WalletPage) OnResume() {
 		}
-
-		pg.listItems = append(pg.listItems, listItem)
+		pg.listLock.Lock()
```
Lock not needed to write to local copy

beansgum commented on ui/page/wallet_page.go:
```diff
@@ -129,9 +132,16 @@ func (pg *WalletPage) ID() string {
 	wallets := pg.WL.SortedWalletList()
 
-	pg.listItems = make([]*walletListItem, 0)
+	pg.listLock.Lock()
```
Lock is not needed here.

beansgum commented on ui/page/wallet_page.go:
```diff
@@ -409,8 +424,14 @@ func (pg *WalletPage) layoutOptionsMenu(gtx layout.Context, optionsMenuIndex int
+
+	return pg.walletsList.Layout(gtx, len(listItems), func(gtx C, i int) D {
+
 		listItem := pg.listItems[i]
```
```suggestion
		listItem := listItems[i]
```

beansgum reviewed (changes requested):
There're some usages of listItems that are not protected by mutex:
https://github.com/planetdecred/godcr/blob/339cdd9bb043cdbe6feb4b8778e022d645f72d56/ui/page/wallet_page.go#L465
https://github.com/planetdecred/godcr/blob/339cdd9bb043cdbe6feb4b8778e022d645f72d56/ui/page/wallet_page.go#L524
https://github.com/planetdecred/godcr/blob/339cdd9bb043cdbe6feb4b8778e022d645f72d56/ui/page/wallet_page.go#L563
https://github.com/planetdecred/godcr/blob/339cdd9bb043cdbe6feb4b8778e022d645f72d56/ui/page/wallet_page.go#L599
https://github.com/planetdecred/godcr/blob/339cdd9bb043cdbe6feb4b8778e022d645f72d56/ui/page/wallet_page.go#L855

Sirmorrison committed:
lock unprotect listitems

beansgum commented on ui/page/wallet_page.go:
```diff
@@ -441,7 +462,10 @@ func (pg *WalletPage) walletSection(gtx layout.Context) layout.Dimensions {
-						return decredmaterial.Clickable(gtx, pg.listItems[i].addAcctClickable, func(gtx C) D {
+						pg.listLock.Lock()
+						listItem := pg.listItems[i]
+						pg.listLock.Unlock()
```
You can use the list item declared earlier here https://github.com/planetdecred/godcr/blob/09cabbcfc0037772251c4648f00dfcf412d6dd81/ui/page/wallet_page.go#L433

beansgum commented on ui/page/wallet_page.go:
```diff
@@ -568,8 +603,11 @@ func (pg *WalletPage) layoutWatchOnlyWallets(gtx layout.Context) D {
 				layout.Rigid(func(gtx C) D {
+					pg.listLock.Lock()
+					listItems := pg.listItems
+					pg.listLock.Unlock()
```
Use the listItems declared earlier https://github.com/planetdecred/godcr/blob/09cabbcfc0037772251c4648f00dfcf412d6dd81/ui/page/wallet_page.go#L566
